modules/catalog-compaction/src/main/java/org/apache/ignite/internal/catalog/compaction/CatalogCompactionRunner.java

```diff
@@ -198,115 +225,127 @@ public CompletableFuture<Boolean> onLowWatermarkChanged(HybridTimestamp newLowWa
         long localMinimum = localMinTimeProvider.time();
 
-        if (catalogByTsNullable(localMinimum) == null) {
-            return CompletableFutures.falseCompletedFuture();
+        if (catalogManagerHelper.catalogByTsNullable(localMinimum) == null) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Catalog compaction skipped, nothing to compact (ts={})", localMinimum);
+            }
+
+            return CompletableFutures.nullCompletedFuture();
         }
 
         return determineGlobalMinimumRequiredTime(topologySnapshot.nodes(), localMinimum)
-                .thenComposeAsync(ts -> {
-                    Catalog catalog = catalogByTsNullable(ts);
+                .thenComposeAsync(timeHolder -> {
+                    long minRequiredTime = timeHolder.minRequiredTime;
+                    HybridTimestamp minActiveTxStartTime = timeHolder.minActiveTxStartTime;
+                    Catalog catalog = catalogManagerHelper.catalogByTsNullable(minRequiredTime);
+
+                    CompletableFuture<Boolean> catalogCompactionFut;
 
                     if (catalog == null) {
-                        return CompletableFutures.falseCompletedFuture();
-                    }
+                        if (LOG.isDebugEnabled()) {
+                            LOG.debug("Catalog compaction skipped, nothing to compact (ts={})", minRequiredTime);
+                        }
 
-                    return requiredNodes(catalog)
-                            .thenCompose(requiredNodes -> {
-                                List<String> missingNodes = missingNodes(requiredNodes, topologySnapshot.nodes());
+                        catalogCompactionFut = CompletableFutures.falseCompletedFuture();
+                    } else {
+                        catalogCompactionFut = tryCompactCatalog(catalog, topologySnapshot).whenComplete((res, ex) -> {
+                            if (ex != null) {
+                                LOG.warn("Catalog compaction has failed (timestamp={})", ex, minRequiredTime);
+                            } else {
+                                if (res && LOG.isDebugEnabled()) {
+                                    LOG.debug("Catalog compaction completed successfully (timestamp={})", minRequiredTime);
+                                } else {
+                                    LOG.debug("Catalog compaction skipped (timestamp={})", minRequiredTime);
+                                }
+                            }
+                        });
+                    }
 
-                                if (!missingNodes.isEmpty()) {
-                                    if (LOG.isDebugEnabled()) {
-                                        LOG.debug("Catalog compaction aborted due to missing cluster members (nodes={})", missingNodes);
-                                    }
+                    if (minActiveTxStartTime == null) {
```

Comment:
I think, we always need to propagate some time. Given that initially partition time is set to min possible clock value, we still must be able to do compaction, even if there is no transactional load. The safest possible option here is, if current node doesn't have active transactions, send current value of low watermark acquired just _before_ collecting time of active transactions